File: linux/umd/src/hal/display_controller.rs
use std::ptr::{addr_of, addr_of_mut};
use std::sync::Arc;
use tokio::task::JoinHandle;
use crate::hal::uio::Uio;
use crate::hal::MemoryMap;

#[repr(C)]
struct DisplayControllerRegisters {
    irq_status: u32,
    irq_mask: u32,
    width: u32,
    height: u32,
    page_addr: u32,
    words: u32,
    ctrl: u32,
}

pub struct DisplayController {
    map: Arc<MemoryMap>,
    draw_done_irq: tokio::sync::watch::Receiver<()>,
    irq_handle: JoinHandle<()>
}

impl DisplayController {
    pub unsafe fn new(mut uio: Uio) -> std::io::Result<Self> {
        let map = Arc::new(uio.map(0)?);

        let (draw_done_tx, draw_done_rx) = tokio::sync::watch::channel(());
        let irq_handle = {
            let map = map.clone();
            tokio::spawn(async move {
                loop {
                    uio.enable_irq();
                    uio.wait_irq().await.expect("wait_irq failed");
                    unsafe {
                        let regs = **map as *mut DisplayControllerRegisters;
                        let status = addr_of!((*regs).irq_status).read_volatile();
                        addr_of_mut!((*regs).irq_status).write_volatile(status);
                    }
                    if draw_done_tx.send(()).is_err() {
                        break;
                    }
                }
            })
        };

        let mut s = Self {
            map,
            draw_done_irq: draw_done_rx,
            irq_handle,
        };

        let size = s.width() * s.height() * 3;
        assert_eq!(size % 8, 0);
        unsafe {
            //disable
            addr_of_mut!((*s.regs_mut()).ctrl).write_volatile(0b0);
            //set buffer size
            addr_of_mut!((*s.regs_mut()).words).write_volatile(size / 8);
            //enable draw done irq only
            addr_of_mut!((*s.regs_mut()).irq_mask).write_volatile(0b10);
        }
        Ok(s)
    }

    pub fn width(&self) -> u32 {
        unsafe {
            addr_of!((*self.regs()).width).read()
        }
    }

    pub fn height(&self) -> u32 {
        unsafe {
            addr_of!((*self.regs()).height).read()
        }
    }

    pub async fn wait_end_of_frame(&mut self) {
        self.draw_done_irq.changed().await.expect("Sender should never be dropped before receiver");
    }

    //TODO: better buffer abstraction instead of passing physical addresses
    pub unsafe fn draw_frame(&mut self, addr: usize) {
        assert_eq!(addr & 0xFFF, 0, "Address must be page alingned");
        let aligned = (addr >> 12) as u32;
        addr_of_mut!((*self.regs_mut()).page_addr).write_volatile(aligned);
        addr_of_mut!((*self.regs_mut()).ctrl).write_volatile(1);
    }

    fn regs(&self) -> *const DisplayControllerRegisters {
        **self.map as *const DisplayControllerRegisters
    }

    fn regs_mut(&mut self) -> *mut DisplayControllerRegisters {
        **self.map as *mut DisplayControllerRegisters
    }
}

impl Drop for DisplayController {
    fn drop(&mut self) {
        self.irq_handle.abort();

        unsafe {
            addr_of_mut!((*self.regs_mut()).ctrl).write_volatile(0);
        }
    }
}

unsafe impl Send for DisplayController {}

File: linux/hell/hal/display_controller.py
import asyncio
from .uio import Uio
from .mmio import u32


__all__ = ["DisplayController"]


IRQ_STATUS = slice(0x00, 0x04)
IRQ_MASK = slice(0x04, 0x08)
WIDTH = slice(0x08, 0x0C)
HEIGHT = slice(0x0C, 0x10)
PAGE_ADDR = slice(0x10, 0x14)
WORDS = slice(0x14, 0x18)
CTRL = slice(0x18, 0x1C)


class DisplayController:
    def __init__(self, uio: Uio):
        self._uio = uio
        self._map = uio.map(0)
        self._draw_done = asyncio.Event()
        asyncio.get_event_loop().create_task(self._handle_irq())

        fb_size = self.width * self.height * 3
        assert fb_size % 8 == 0
        self._map[CTRL] = u32(0)
        self._map[WORDS] = u32(fb_size // 8)
        self._map[IRQ_MASK] = u32(0b10)

    def __del__(self):
        self._map[CTRL] = u32(0)

    @property
    def width(self) -> int:
        return u32(self._map[WIDTH])

    @property
    def height(self) -> int:
        return u32(self._map[HEIGHT])

    async def wait_end_of_frame(self):
        await self._draw_done.wait()
        self._draw_done.clear()

    def draw_frame(self, addr: int):
        assert addr & 0xFFF == 0
        self._map[PAGE_ADDR] = u32(addr >> 12)
        self._map[CTRL] = u32(1)

    async def _handle_irq(self):
        while True:
            self._uio.enable_irq()
            await self._uio.wait_irq()

            irq_status = u32(self._map[IRQ_STATUS])
            self._map[IRQ_STATUS] = u32(irq_status)

            self._draw_done.set()
